src/1-theory/10-async.adoc

```diff
@@ -0,0 +1,407 @@
+network caching can help a bit here, but even then it sucks the battery life on
+mobile devices.
+
+Generally short polling is rubbish, and you should try to avoid it.
+
+=== Long Polling
+
+A slight tweak on the concept of short polling, long polling maintains an open
+connection with the server until the answer comes in.
+
+[quote,sp3c1,https://stackoverflow.com/questions/4642598/short-polling-vs-long-polling-for-real-time-web-applications]
+----
+00:00:00 C-> Is the cake ready?
+00:00:03 S-> Yeah. Have some lad.
+00:00:03 C-> Is the other cake ready?
+----
+
+See here that instead of answering immediately, the server waited 3 seconds
+before responding with a single update.
+
+Generally folks recommend long polling when the client is only waiting for one
+answer, so if they are ordering a single cake and it might be ready soon, then
+great, let that thread focus on replying about the one cake, and when that cake
+is doing the connection is closed. If the client wants multiple status updates,
+or is interested in multiple cakes, this might not be the way to go.
+
+Personal experiences tell me this sort of blocking activity can lead to quite a
+mess when you have more than a smattering of users. Blocking an entire web
+thread for 3 seconds can certainly be a problem when you have a lot of clients
+asking about a lot of cakes.
+
+For example, in Rails land this is a pretty common server config:
+
+[source,ruby]
+----
+workers Integer(ENV['WEB_CONCURRENCY'] || 2)
+threads_count = Integer(ENV['RAILS_MAX_THREADS'] || 5)
+threads threads_count, threads_count
+
+rackup      DefaultRackup
+port        ENV['PORT']     || 3000
+environment ENV['RACK_ENV'] || 'development'
+----
+
+If there was only one dyno running (a.k.a instance, pod, whatever) then 2 * 5 =
+10 and thats only 10 cake orders happening at the same time and you've run out of
+threads for other things, like accepting payments, or registering a user.
+
+Even having a slightly bulkier setup, with more dynos, and more processes, and more threads,
+its simply a numbers game where you are throwing more money at the situation.
+
+=== Web Hooks
+
+Both these polling approaches are essentially a pull model, so lets look at a
+push model: Web Hooks. Clients register a "callback URL" through a
+GUI or API, and they may also specify specific types of information they are
+interested in receiving. When relevant events happen on the server, it will fire
+a payload at that URL, which usually takes the form of a HTTP
+POST request.
+
+.Examples of Web Hooks in popular APIs.
+- https://api.slack.com/incoming-webhooks[Slack API > Incoming Webhooks]
+- https://stripe.com/docs/webhooks[Stripe API > Webhooks]
+
+A few years ago my friend and I made a website that sold silly programmer joke
+t-shirts. We created the whole thing over beers, and we made some whopping
+mistakes. Our payment acceptance sequence looked like the synchronous one above, as
```

Comment:
> looked like the synchronous one above

Not sure what you're referring to here?

Reply:
Ah, this used to be one flow with the diagrams mixed into the definitiuons. Then I split the definitions out first and did the diagrams afterwards as a use case. Fixed!